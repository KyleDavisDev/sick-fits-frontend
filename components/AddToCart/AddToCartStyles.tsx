import styled from "styled-components";

export const StyledButton = styled.button`
  &:hover {
    cursor: pointer;
    background-color: ${props => props.theme.smoke};
  }
`;
