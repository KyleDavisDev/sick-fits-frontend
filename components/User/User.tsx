import * as React from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import { ReactElement } from "react";

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      permissions
      cart {
        created
        id
        items {
          id
          quantity
          item {
            id
            price
            image
            title
            description
          }
        }
      }
    }
  }
`;

export interface IUserProps {
  children(data: any): ReactElement;
}

const User: React.FunctionComponent<IUserProps> = props => {
  return (
    <Query {...props} query={CURRENT_USER_QUERY}>
      {payload => {
        return props.children(payload);
      }}
    </Query>
  );
};

export default User;
