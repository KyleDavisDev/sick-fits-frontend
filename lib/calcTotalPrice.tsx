export default function(items) {
  return items.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity * cartItem.item.price;
  }, 0);
}
