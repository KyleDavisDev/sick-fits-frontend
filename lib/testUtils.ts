import casual from "casual";
import { ReactWrapper } from "enzyme";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  created: casual.unix_time,
  id: casual.uuid,
  price: 5000,
  user: null,
  image: "dog-small.jpg",
  title: "dogs are best",
  description: "dogs",
  largeImage: "dog.jpg"
});

const fakeUser = () => ({
  __typename: "User",
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: []
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: `${casual.word}.jpg`,
  title: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words()
});

const fakeOrder = () => ({
  __typename: "Order",
  id: casual.uuid,
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2018-04 - 06T19: 24: 16.000Z",
  user: fakeUser()
});

const fakeCartItem = (overrides?: any) => ({
  __typename: "CartItem",
  id: casual.uuid,
  created: casual.unix_time,
  quantity: casual.integer(0, 15),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides
});

const fakeCart = (overrides?: any) => ({
  __typename: "Cart",
  id: casual.uuid,
  items: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
  user: fakeUser(),
  created: casual.unix_time,
  updated: casual.unix_time,
  isActive: true,
  ...overrides
});

// Fake LocalStorage
class LocalStorageMock {
  store: any;
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

export {
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeCart,
  fakeOrder,
  fakeOrderItem,
  simulateInputChange
};
