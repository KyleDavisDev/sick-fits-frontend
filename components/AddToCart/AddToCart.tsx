import * as React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import { StyledButton } from "./AddToCartStyles";
import { CURRENT_USER_QUERY } from "../User/User";

export const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
    }
  }
`;

interface IAddToCartProps {
  id: string;
}

class AddToCart extends React.Component<IAddToCartProps> {
  render() {
    const { id } = this.props;
    return (
      <Mutation
        mutation={ADD_TO_CART_MUTATION}
        variables={{ id }}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      >
        {addToCart => {
          return <StyledButton onClick={addToCart}>Add to Cart</StyledButton>;
        }}
      </Mutation>
    );
  }
}

export default AddToCart;
