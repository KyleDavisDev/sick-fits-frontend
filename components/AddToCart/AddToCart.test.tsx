import "jsdom-global/register";
import * as React from "react";
import { mount } from "enzyme";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { ApolloConsumer } from "react-apollo";
import AddToCart, { ADD_TO_CART_MUTATION } from "./AddToCart";
import { CURRENT_USER_QUERY } from "../User/User";
import { fakeUser, fakeItem, fakeCart } from "../../lib/testUtils";

const me = fakeUser();
const cart = fakeCart();
const item = fakeItem();
const mocks = [
  {
    request: { query: ADD_TO_CART_MUTATION },
    result: {
      data: { addToCart: { __typename: "Cart", id: item.id } }
    }
  },
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: { me: { ...me, cart: { ...cart, items: [...cart.items, item] } } }
    }
  }
];

describe("<AddToCart />", () => {
  it("renders and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <AddToCart id={item.id} />
      </MockedProvider>
    );

    await wait();
    wrapper.update();

    // check render
    expect(wrapper).toBeTruthy();

    // check snaps
    expect(wrapper.find("button")).toMatchSnapshot();

    console.log(wrapper.debug());
  });
});
