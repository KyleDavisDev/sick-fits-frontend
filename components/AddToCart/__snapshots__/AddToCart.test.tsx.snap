// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart /> renders and matches snapshot 1`] = `
<button
  className="sc-bdVaJa dxkKyp"
  disabled={false}
  onClick={[Function]}
>
  Add
   to Cart
</button>
`;
